coderd/rbac/policy.rego

```diff
@@ -19,121 +19,98 @@ bool_flip(b) = flipped {
+default site = 0
+site := num {
+	allow := { x |
+    	perm := input.subject.roles[_].site[_]
+        perm.action in [input.action, "*"]
+		perm.resource_type in [input.object.type, "*"]
+        x := bool_flip(perm.negate)
     }
+    num := number(allow)
 }
 
-# If a user is not part of an organization, 'org_non_member' is set to true
-org_non_member {
-    input.object.org_owner != ""
-    # Identify if the user is in the org
-    roles := input.subject.roles
-    every role in roles {
-        not role.org[input.object.org_owner]
-    }
+org_members := { orgID |
+	input.subject.roles[_].org[orgID]
 }
 
-# org is two rules that equate to the following
-#	if org_non_member { return {false} }
-#	else { org_member }
-#
-# It is important both rules cannot be true, as the `org` rules cannot produce multiple outputs.
-default org = {}
-org = set {
-    # We have to do !org_non_member because rego rules must evaluate to 'true'
-    # to have a value set.
-    # So we do "not not-org-member" which means "subject is in org"
-    not org_non_member
-    set = org_member
+default org = 0
+org := num {
+	orgPerms := { id: num |
+		id := org_members[_]
+		set := { x |
+			perm := input.subject.roles[_].org[id][_]
+			perm.action in [input.action, "*"]
+			perm.resource_type in [input.object.type, "*"]
+			x := bool_flip(perm.negate)
+		}
+		num := number(set)
+	}
+
+	num := orgPerms[input.object.org_owner]
 }
 
-org = set {
-    org_non_member
-    set = {false}
+# 'org_mem' is set to true if the user is an org member
+# or if the object has no org.
+org_mem := true {
+	input.object.org_owner != ""
+	input.object.org_owner in org_members
 }
 
-# The allow block is quite simple. Any set with `false` cascades down in levels.
-# Authorization looks for any `allow` statement that is true. Multiple can be true!
-# Note that the absence of `allow` means "unauthorized".
-# An explicit `"allow": true` is required.
-
-# site allow
-allow {
-    # No site wide deny
-    not false in site
-    # And all permissions are positive
-    site[_]
+org_mem := true {
+	input.object.org_owner == ""
 }
 
-# OR
+default user = 0
+user := num {
+    input.object.owner != ""
+    input.subject.id = input.object.owner
+	allow := { x |
+    	perm := input.subject.roles[_].user[_]
+        perm.action in [input.action, "*"]
+		perm.resource_type in [input.object.type, "*"]
+        x := bool_flip(perm.negate)
+    }
+    num := number(allow)
+}
 
-# org allow
+default allow = false
+# Site
 allow {
-    # No site or org deny
-    not false in site
-    not false in org
-    # And all permissions are positive
-    org[_]
+	site = 1
 }
 
-# OR
+# Org
+allow {
+	not site = -1
+	org = 1
+}
 
-# user allow
+# User
 allow {
-    # No site, org, or user deny
-    not false in site
-    not false in org
-    not false in user
-    # And all permissions are positive
-    user[_]
+	not site = -1
+	not org = -1
+	org_mem
```

Comment:
I'm curious about why this `org_mem` clause is required.  `user` will only be 1 if the subject owns the object.  Is it possible for us to have an object that is owned by a user and org, such that the user is not a member of the org?

If it were a workspace, that workspace would essentially be orphaned because the owner would no longer have access to it.

Reply:
This is correct. If you kick a user from an org, those workspaces they own in that org are orphaned from the user. This is _strange_, but that is the documented behavior.